src/promise.rs

```diff
@@ -0,0 +1,338 @@
+            *touched_clone.borrow_mut() = true;
+        });
+
+        promise.fulfill(10);
+        assert_eq!(true, touched.take())
+    }
+
+    #[test]
+    fn test_all_of_success() {
+        let touched = Rc::new(RefCell::new(false));
+        let touched_clone = touched.clone();
+
+        let promise1 = Promise::<i32>::new();
+        let promise2 = Promise::<i32>::new();
+
+        let all_promise = Promise::all_of(vec![promise1.clone(), promise2.clone()]);
+
+        promise1.fulfill(42);
+        promise2.fulfill(100);
+
+        all_promise
+            .then(move |results| {
+                assert_eq!(results.len(), 2);
+                assert_eq!(results[0], 42);
+                assert_eq!(results[1], 100);
+                *touched_clone.borrow_mut() = true;
+            })
+            .catch(|_err| {
+                panic!("Should not reach here");
+            });
+
+        assert_eq!(true, touched.take())
+    }
+
+    #[test]
+    fn test_all_of_failure() {
+        let touched = Rc::new(RefCell::new(false));
+        let touched_clone = touched.clone();
+
+        let promise1 = Promise::<i32>::new();
+        let promise2 = Promise::<i32>::new();
+
+        let all_promise = Promise::all_of(vec![promise1.clone(), promise2.clone()]);
+
+        promise1.reject("Error 1".to_string());
+        promise2.reject("Error 2".to_string());
+
+        all_promise
+            .then(|_results| {
+                panic!("Should not reach here");
+            })
+            .catch(move |err| {
+                assert_eq!(err, "Error 1");
+                *touched_clone.borrow_mut() = true;
+            });
+
+        assert_eq!(true, touched.take())
+    }
+
+    #[test]
+    fn test_all_of_mixed_results() {
+        let touched = Rc::new(RefCell::new(false));
+        let touched_clone = touched.clone();
+
+        let promise1 = Promise::<i32>::new();
+        let promise2 = Promise::<i32>::new();
+
+        let all_promise = Promise::all_of(vec![promise1.clone(), promise2.clone()]);
+
+        promise1.reject("Error".to_string());
+        promise2.fulfill(100);
+
+        all_promise
+            .then(|_| {
+                panic!("Should not reach here");
+            })
+            .catch(move |reason| {
+                assert_eq!(reason, "Error".to_string());
+                *touched_clone.borrow_mut() = true;
+            });
+
+        assert_eq!(true, touched.take())
+    }
+
+    #[test]
+    fn test_all_of_empty() {
+        let touched = Rc::new(RefCell::new(false));
+        let touched_clone = touched.clone();
+
+        let all_promise = Promise::<i32>::all_of(vec![]);
+
+        all_promise
+            .then(move |results| {
+                assert!(results.is_empty());
+                *touched_clone.borrow_mut() = true;
+            })
+            .catch(|_err| {
+                panic!("Should not reach here");
+            });
+
+        assert_eq!(true, touched.take())
+    }
+}
```

Comment:
This makes me a bit nervous. I don't think this SDK should be in the business of maintaining a generic async machinery like that... However, I've never liked the existing flow, and I've been looking into adding something like this myself over the last week (for sync/async KV store), so this is definitely a welcomed feature. Thanks!

I'm dealing with something else right now, but I'll try to review it later this week.